hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -1021,41 +1033,52 @@ void doDrain(final List<RAMQueueEntry> entries) throws InterruptedException {
 
   private void persistToFile() throws IOException {
     assert !cacheEnabled;
-    FileOutputStream fos = null;
-    ObjectOutputStream oos = null;
-    try {
+    try (ObjectOutputStream oos = new ObjectOutputStream(
+      new FileOutputStream(persistencePath, false))){
       if (!ioEngine.isPersistent()) {
         throw new IOException("Attempt to persist non-persistent cache mappings!");
       }
-      fos = new FileOutputStream(persistencePath, false);
-      oos = new ObjectOutputStream(fos);
+      oos.write(ProtobufUtil.PB_MAGIC);
+      byte[] checksum = ((PersistentIOEngine) ioEngine).calculateChecksum();
+      oos.writeInt(checksum.length);
+      oos.write(checksum);
       oos.writeLong(cacheCapacity);
       oos.writeUTF(ioEngine.getClass().getName());
       oos.writeUTF(backingMap.getClass().getName());
       oos.writeObject(deserialiserMap);
       oos.writeObject(backingMap);
-    } finally {
-      if (oos != null) oos.close();
-      if (fos != null) fos.close();
+    } catch (NoSuchAlgorithmException e) {
+      LOG.error("No such algorithm : " + algorithm + "! Failed to persist data on exit",e);
     }
   }
 
   @SuppressWarnings("unchecked")
-  private void retrieveFromFile(int[] bucketSizes) throws IOException, BucketAllocatorException,
+  private void retrieveFromFile(int[] bucketSizes) throws IOException,
       ClassNotFoundException {
     File persistenceFile = new File(persistencePath);
     if (!persistenceFile.exists()) {
       return;
     }
     assert !cacheEnabled;
-    FileInputStream fis = null;
     ObjectInputStream ois = null;
     try {
       if (!ioEngine.isPersistent())
         throw new IOException(
             "Attempt to restore non-persistent cache mappings!");
-      fis = new FileInputStream(persistencePath);
-      ois = new ObjectInputStream(fis);
+      ois = new ObjectInputStream(new FileInputStream(persistencePath));
+      // there are two situations that can cause verification to fail
+      if (!((FileIOEngine) ioEngine).verifyFileIntegrity(persistencePath, ois)) {
+        // 1. persistent file is old version, it's not support verification,
+        // so reopen ObjectInputStream and read the persistent file from head
+        if (((FileIOEngine) ioEngine).isOldVersion()) {
+          ois.close();
+          ois = new ObjectInputStream(new FileInputStream(persistencePath));
+        } else {
+          // 2. there are some IOException like file not exist, and the FileIOEngine
+          // has been reinitialized
+          return;
```

Comment:
here we return, what about backingMap?

Reply:
In "finally", the persistentFile(or backingMap) will be deleted.